pkg/aks/create.go

```diff
@@ -23,17 +23,134 @@ func CreateResourceGroup(ctx context.Context, groupsClient *resources.GroupsClie
 	return err
 }
 
-// CreateOrUpdateCluster creates a new managed Kubernetes cluster
-func CreateOrUpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
+// CreateCluster creates a new managed Kubernetes cluster. In this case, there will be no existing upstream cluster.
+// We are provisioning a brand new one.
+func CreateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
 	spec *aksv1.AKSClusterConfigSpec, phase string) error {
 
+	managedCluster, err := CreateManagedCluster(ctx, cred, spec, phase)
+	if err != nil {
+		return err
+	}
+
+	_, err = clusterClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		spec.ClusterName,
+		managedCluster,
+	)
+
+	return err
+}
+
+// UpdateCluster updates an existing managed Kubernetes cluster. Before updating, it pulls any existing configuration
+// and then only updates managed fields.
+func UpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
+	spec *aksv1.AKSClusterConfigSpec, phase string) error {
+
+	// Create a new managed cluster from the AKS cluster config
+	managedCluster, err := CreateManagedCluster(ctx, cred, spec, phase)
+	if err != nil {
+		return err
+	}
+
+	// Pull the upstream cluster state. If the upstream cluster couldn't be found (not expected), then create a new
+	// cluster. If there was a bad error, return it.
+	upstreamCluster, err := clusterClient.Get(ctx, spec.ResourceGroup, spec.ClusterName)
+	if err != nil {
+		if strings.Contains(err.Error(), spec.ClusterName + " was not found") {
+			return CreateCluster(ctx, cred, clusterClient, spec, phase)
+		} else {
+			return err
+		}
+	}
+
+	// Upstream cluster state was successfully pulled. Merge in updates without overwriting upstream fields: we never
+	// want to overwrite preconfigured values in Azure with nil values. So if a field on the managed cluster is nil,
+	// set it to the upstream value.
```

Comment:
It looks to me that the current approach is "build the AKS cluster object based on everything Rancher knows about (`managedCluster`) and set all the fields Rancher doesn't know about from object we got from AKS (`upstreamCluster`)". I think flipping this around makes more sense - get current object from AKS and update all the fields Rancher knows about on it before updating.
The big downside of the current approach is added maintenance - any time new (top-level) AKS field is added, we need to update this logic. I think it's much more natural to handle what Rancher supports ("knows about") instead of the reverse ("we don't know how much we don't know" type of situation).

Reply:
Ahh yes.. on second thought I'm with you here. I ran into some minor complications trying to figure out which upstream fields in Azure I should check for on the AKS cluster object (`managedCluster`). If we do it that we, we won't have to monitor a list of upstream fields that MAY be set.